zebra-state/src/service/check/tests/utxo.rs

```diff
@@ -0,0 +1,663 @@
+//! Randomised property tests for UTXO contextual validation
+
+use std::{convert::TryInto, sync::Arc};
+
+use proptest::prelude::*;
+
+use zebra_chain::{
+    block::{Block, Height},
+    fmt::TypeNameToDebug,
+    serialization::ZcashDeserializeInto,
+    transaction::{LockTime, Transaction},
+    transparent,
+};
+
+use crate::{
+    service::{
+        arbitrary::{new_state_with_mainnet_genesis, transaction_v4_from_coinbase},
+        StateService,
+    },
+    tests::Prepare,
+    FinalizedBlock,
+    ValidateContextError::{DuplicateTransparentSpend, MissingTransparentOutput},
+};
+
+// These tests use the `Arbitrary` trait to easily generate complex types,
+// then modify those types to cause an error (or to ensure success).
+//
+// We could use mainnet or testnet blocks in these tests,
+// but the differences shouldn't matter,
+// because we're only interested in spend validation,
+// (and passing various other state checks).
+
+proptest! {
+    /// Make sure an arbitrary transparent spend from a previous transaction in this block
+    /// is accepted by state contextual validation.
+    ///
+    /// This test makes sure there are no spurious rejections that might hide bugs in the other tests.
+    /// (And that the test infrastructure generally works.)
+    ///
+    /// It also covers a potential edge case where later transactions can spend outputs
+    /// of previous transactions in a block, but earlier transactions can not spend later outputs.
+    #[test]
+    fn accept_arbitrary_transparent_spend_from_this_block(
```

Comment:
Minor: would it make sense to run `rustfmt` inside the `proptest!` block? What I usually do is replace `proptest!` with `impl Proptest`, and comment all function parameters (because they have a `$pat in $expr` custom syntax). That makes `rustfmt` runnable, but then I have to undo the parameter and `impl Proptest` changes afterwards :laughing:

Reply:
Could get messy! Let's do that as the last commit in this PR, if we remember.

Reply:
Actually it was really messy, so I did the nullifier and utxo tests in PR #2515.